### PR TITLE
Support alpha beta scaling for GEMM

### PR DESCRIPTION
**Pull request by rocking5566**
Support C = (alpha * A * B) + (beta * C0)

### CODE REVIEW

File: composable_kernel/include/tensor_operation/element_wise_operation.hpp [hunk 1]
```diff
@@ -12,6 +12,31 @@ struct PassThrough
...
+        dst = static_cast<T1>(src_y + static_cast<T2>(bias));
+    }
+};
+
+struct AlphaBetaAdd
+{
+    AlphaBetaAdd(float alpha, float beta) : alpha_(alpha), beta_(beta) {}
+
+    template <typename T1, typename T2, typename T3>
+    __host__ __device__ constexpr void operator()(T1& dst, const T2& src_y, const T3& bias) const
```
Comment (asroy) [inline composable_kernel/include/tensor_operation/element_wise_operation.hpp:29]: Let's not use templated type, but real type here

File: composable_kernel/include/tensor_operation/element_wise_operation.hpp [hunk 2]
```diff
@@ -12,6 +12,31 @@ struct PassThrough
     __host__ __device__ void operator()(half_t& y, const half_t& x) const { y = x; }
 };
 
+struct Add
+{
+    template <typename T1, typename T2, typename T3>
```
Comment (asroy) [inline composable_kernel/include/tensor_operation/element_wise_operation.hpp:17]: Let's not use templated type, but real type here

File: composable_kernel/include/tensor_operation/element_wise_operation.hpp (hunk 1 above)
Comment (rocking5566) [reply composable_kernel/include/tensor_operation/element_wise_operation.hpp]: Fix
https://github.com/ROCmSoftwarePlatform/composable_kernel/commit/9bafd938f7cf09a809c0183074cf30f8664cd93b#diff-990561d35152da39a727f324616ba178793b8b1e20ecd17194b48988282cbd66R36-R47

File: composable_kernel/include/tensor_operation/element_wise_operation.hpp (hunk 2 above)
Comment (rocking5566) [reply composable_kernel/include/tensor_operation/element_wise_operation.hpp]: Fix
https://github.com/ROCmSoftwarePlatform/composable_kernel/commit/9bafd938f7cf09a809c0183074cf30f8664cd93b#diff-990561d35152da39a727f324616ba178793b8b1e20ecd17194b48988282cbd66R15-R30

File: composable_kernel/include/tensor_operation/element_wise_operation.hpp [hunk 3]
```diff
@@ -12,6 +12,45 @@ struct PassThrough
     __host__ __device__ void operator()(half_t& y, const half_t& x) const { y = x; }
 };
 
+struct Add
+{
+    __host__ __device__ constexpr void
+    operator()(float& dst, const float& src_y, const float& bias) const
```
Comment (asroy) [inline composable_kernel/include/tensor_operation/element_wise_operation.hpp:18]: Just use y for output and x for input. Word like "bias" only has meaning in GEMM+Bias fusion, but this "Add" operator will be use in many other places.

operator()(float& y, const float& x0, const float& x1) const

**Review by asroy — DISMISSED** [review]

File: composable_kernel/include/tensor_operation/element_wise_operation.hpp (hunk 3 above)
Comment (rocking5566) [reply composable_kernel/include/tensor_operation/element_wise_operation.hpp]: https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/78/commits/8a896c4516fe556d043ff39c2ec2e7436d73f3bc

File: composable_kernel/include/tensor_operation/element_wise_operation.hpp [hunk 4]
```diff
@@ -12,6 +12,45 @@ struct PassThrough
...
+    {
+        // FIXME - Use float (acc type) bias in the future.
```
[the remark this answers is not in the document]
Comment (rocking5566) [reply composable_kernel/include/tensor_operation/element_wise_operation.hpp:19]: Fix in https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/78/commits/8a896c4516fe556d043ff39c2ec2e7436d73f3bc

File: composable_kernel/include/tensor_operation/element_wise_operation.hpp (hunk 3 above)
Comment (rocking5566) [reply composable_kernel/include/tensor_operation/element_wise_operation.hpp]: Fix in https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/78/commits/8a896c4516fe556d043ff39c2ec2e7436d73f3bc

File: composable_kernel/include/tensor_operation/element_wise_operation.hpp [hunk 5]
```diff
@@ -12,6 +12,45 @@ struct PassThrough
...
+        dst = src_y + bias;
+    }
+
+    __host__ __device__ constexpr void
+    operator()(half_t& dst, const half_t& src_y, const half_t& bias) const
```
[the remark this answers is not in the document]
Comment (rocking5566) [reply composable_kernel/include/tensor_operation/element_wise_operation.hpp]: Fix in https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/78/commits/8a896c4516fe556d043ff39c2ec2e7436d73f3bc

File: composable_kernel/include/tensor_operation/element_wise_operation.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (rocking5566) [reply composable_kernel/include/tensor_operation/element_wise_operation.hpp]: Remove in https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/78/commits/10560a7166240b7ff08f301696102df59efba4e2

File: device_operation/include/device_gemm_xdl_c_shuffle_bias_2d.hpp [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+              N01_{N01},
+              a_element_op_{a_element_op},
+              b_element_op_{b_element_op},
+              c_element_op_{c_element_op}
+        {
+            a_grid_desc_k0_m_k1_ =
+                DeviceGemmXdl_C_Shuffle_Bias_2d::MakeAGridDescriptor_K0_M_K1(M, K, StrideA);
+            b_grid_desc_k0_n_k1_ =
+                DeviceGemmXdl_C_Shuffle_Bias_2d::MakeBGridDescriptor_K0_N_K1(K, N, StrideB);
+            c0_grid_desc_m_n_ =
```
Comment (asroy) [inline device_operation/include/device_gemm_xdl_c_shuffle_bias_2d.hpp:220]: bias could be using a different stride from C matrix.

Please fix it in future PR.

File: device_operation/include/device_gemm_xdl_c_shuffle_bias_2d.hpp [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+    typename BBlockTransferSrcAccessOrder,
+    ck::index_t BBlockTransferSrcVectorDim,
+    ck::index_t BBlockTransferSrcScalarPerVector,
+    ck::index_t BBlockTransferDstScalarPerVector_K1,
+    bool BBlockLdsAddExtraN,
+    index_t CShuffleMXdlPerWavePerShuffle,
+    index_t CShuffleNXdlPerWavePerShuffle,
+    typename CBlockTransferClusterLengths_MBlock_MXdlPerWave_MWaveMPerXdl_NBlock_NXdlPerWave_NWaveNPerXdl,
+    index_t CBlockTransferScalarPerVector_NWaveNPerXdl>
+struct DeviceGemmXdl_C_Shuffle_Bias_2d
```
Comment (asroy) [inline device_operation/include/device_gemm_xdl_c_shuffle_bias_2d.hpp:58]: not sure we should call it "bias 2d", is there a better name?

**Review by asroy — APPROVED** [review]

**Pull request merged**